mediacloud/mediawords/util/web/user_agent/throttled.py

```diff
@@ -0,0 +1,67 @@
+"""Implement ThrottledUserAgent as a sub class of mediawords.util.web.UserAgent with per domain throttling."""
+
+import typing
+
+import mediawords.db
+import mediawords.util.config
+from mediawords.util.web.user_agent import UserAgent
+from mediawords.util.web.user_agent.request.request import Request
+from mediawords.util.web.user_agent.response.response import Response
+
+from mediawords.util.log import create_logger
+log = create_logger(__name__)
+
+# default amount of time in between requests
+_DEFAULT_DOMAIN_TIMEOUT = 10
+
+
+class McThrottledUserAgentTimeoutException(Exception):
+    """Exception raised when a ThrottledUserAgent request fails to get a domain request lock."""
+
+    pass
+
+
+class ThrottledUserAgent(UserAgent):
+    """Add per domain throttling to mediawords.util.web.UserAgent."""
+
+    def __init__(self, db: mediawords.db.DatabaseHandler, domain_timeout: typing.Optional[int]=None) -> None:
+        """
+        Add database handler and domain_timeout to UserAgent object.
+
+        If domain_timeout is not specified, use mediawords.throttles_user_agent_domain_timeout from mediawords.yml.
+        If not present in mediawords.yml, use _DEFAULT_DOMAIN_TIMEOUT.
+        """
+        self.db = db
+        self.domain_timeout = domain_timeout
+
+        if self.domain_timeout is None:
+            config = mediawords.util.config.get_config()
+            if 'throttled_user_agent_domain_timeout' in config['mediawords']:
+                self.domain_timeout = int(config['mediawords']['throttled_user_agent_domain_timeout'])
+            if self.domain_timeout is None or self.domain_timeout < 1:
+                self.domain_timeout = _DEFAULT_DOMAIN_TIMEOUT
+
+        UserAgent.__init__(self)
```

Comment:
should probably be `super().__init__()` (unless you're sure this is what you want - let me know if you want elaboration)

Reply:
fixed.